File: python/longest_prefix.py
# input strings/output strings
# for i in strs, 取第一個字母, 比較其他字母有無一樣, 有的話存到新的box
# 問題點  怎麼取1st 字母

from typing import List

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            return ""


        shortest = min(strs, key=len)
        for i, char in enumerate(shortest):
            for s in strs:
                if s[i] != char:
                    return shortest[:i]

        return shortest


#test
test = ["lisa", "lily", "liar"]
result = Solution().longestCommonPrefix(test)
print(result)

File: rust/longest_prefix.rs
struct Solution;
impl Solution {
    pub fn longest_common_prefix(strs: Vec<String>) -> String {
        if strs.is_empty() {
            return "".to_string();
        }       

        let min_len = strs.iter().map(|a| a.len()).min().unwrap();
        let mut prefix = String::new();

        for i in 0..min_len {
            let c = strs[0].as_bytes()[i];

            if strs.iter().any(|s| s.as_bytes()[i] != c) {
                break;
            }
            prefix.push(c as char);
        }
        prefix
    }
}


fn main() {
    let test:Vec<String> = ["lisa", "lily", "libra"].into_iter().map(String::from).collect();
    let result = Solution::longest_common_prefix(test);
    println!("{}", result)
}
